karapace/schema_reader.py

```diff
@@ -250,109 +262,131 @@ def handle_messages(self):
-                self.log.info("Deleting subject: %r, version: %r", subject, value["version"])
-                if not value["version"] in self.subjects[subject]["schemas"]:
-                    self.schemas[value["id"]] = typed_schema
-                else:
-                    self.subjects[subject]["schemas"][value["version"]]["deleted"] = True
-            elif value.get("deleted", False) is False:
-                self.log.info("Adding new version of subject: %r, value: %r", subject, value)
-                self.subjects[subject]["schemas"][value["version"]] = {
-                    "schema": typed_schema,
-                    "version": value["version"],
-                    "id": value["id"],
-                    "deleted": value.get("deleted", False),
-                }
-                self.log.info("Setting schema_id: %r with schema: %r", value["id"], value["schema"])
-                with self.id_lock:
-                    self.schemas[value["id"]] = typed_schema
-                if value["id"] > self.global_schema_id:  # Not an existing schema
-                    self.global_schema_id = value["id"]
+                self.log.info("Setting subject: %r config to: %r, value: %r", subject, value["compatibilityLevel"], value)
+                self.subjects[subject]["compatibility"] = value["compatibilityLevel"]
+        elif value is not None:
+            self.log.info("Setting global config to: %r, value: %r", value["compatibilityLevel"], value)
+            self.config["compatibility"] = value["compatibilityLevel"]
+
+    def _handle_msg_delete_subject(self, key: dict, value: Optional[dict]) -> None:  # pylint: disable=unused-argument
+        if value is None:
+            self.log.error("DELETE_SUBJECT record doesnt have a value, should have")
+            return
+
+        subject = value["subject"]
+        if subject not in self.subjects:
+            self.log.error("Subject: %r did not exist, should have", subject)
+        else:
+            self.log.info("Deleting subject: %r, value: %r", subject, value)
+            updated_schemas = {
+                key: self._delete_schema_below_version(schema, value["version"])
+                for key, schema in self.subjects[subject]["schemas"].items()
+            }
+            self.subjects[value["subject"]]["schemas"] = updated_schemas
+
+    def _handle_msg_schema_hard_delete(self, key: dict) -> None:
+        subject, version = key["subject"], key["version"]
+
+        if subject not in self.subjects:
+            self.log.error("Hard delete: Subject %s did not exist, should have", subject)
+        elif version not in self.subjects[subject]["schemas"]:
+            self.log.error("Hard delete: Version %d for subject %s did not exist, should have", version, subject)
+        else:
+            self.log.info("Hard delete: subject: %r version: %r", subject, version)
+            self.subjects[subject]["schemas"].pop(version, None)
```

Comment:
Here is where the memory leak mentioned above is still a problem:

https://github.com/aiven/karapace/pull/375/files#r839821293

The previous fix was to use a WeakDict: 

https://github.com/aiven/karapace/pull/284/files#diff-d0d95f66bd9730341ddb7c386650f61768378934401a70d05ec9dac690634ea8R126